storage/replica.go

```diff
@@ -1421,20 +1421,25 @@ func (r *Replica) handleRaftReady() error {
 	batch := r.store.Engine().NewBatch()
 	defer batch.Close()
 	if !raft.IsEmptySnap(rd.Snapshot) {
+		// TODO(peter): Could use a "distinct" batch within applySnapshot, but we
+		// also need the Batch.Defer mechanism.
 		var err error
 		if lastIndex, err = r.applySnapshot(batch, rd.Snapshot); err != nil {
 			return err
 		}
 		// TODO(bdarnell): update coalesced heartbeat mapping with snapshot info.
 	}
+	// We know that all of the writes from here forward will be to distinct keys.
```

Comment:
Isn't the right thing to require here that none of the keys written have previously been mutated in the batch? Reads get passed through to the underlying engine (not the batch). We definitely mutate a lot of keys when we apply a snapshot, conceivably the hard state, last index, ... (maybe we don't currently, but putting this `Distinct()` business in for minor gains seems like a good recipe for a future two-day debugging session).